### ISSUE TITLE
Is there a way to generate the number of TP/TN/FP/FN for each test image using the detect.py script?

### ISSUE THREAD

**Issue by ib124** [ultralytics/yolov5#5713]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

I am doing some detection accuracy analysis, and I am looking to model how different confidence settings (i.e. ```--conf 0.6``` affect the number of true positive/false positive detections for my data. Is there any way the ```detect.py``` script can be modified to list the TP/FP/TN/FN values for each class of each image? I have a custom model that has multiple classes trained, but I only want these values for one class.

Note: I know that the ```val.py``` script graphs metrics such as the F1, Precision-Recall curve, etc., but I'm just trying to get some of the raw values for my individual calculations.

**Comment by glenn-jocher** [comment]
@ib124 no of course not. Where do you expect TP values to be produced in detect.py exactly? Where are the labels used in your imaginary detect.py coming from?

**Comment by ib124** [comment]
@glenn-jocher In that case, is there a way to print out these values using the ```val.py``` script?

**Comment by glenn-jocher** [comment]
@ib124 yes, that's a possibility!

Several users have been asking for this but we don't have it enabled by default. You can access these values directly in the code here, there is one FP/TP vector per IoU threshold 0.5:0.05:0.95:
https://github.com/ultralytics/yolov5/blob/eb51ffdcac466e553607c470b0e8f19d5a61da67/utils/metrics.py#L54-L56

**Comment by ib124** [comment]
That is what I needed. Thank you!

**Comment by glenn-jocher** [comment]
@ib124 another reason we don't display TP and FP is because the information displayed is a sufficient statistic to reconstruct these, so displaying these would be redundant. Anyone can reconstruct these using the provided metrics, same with F1. See https://en.wikipedia.org/wiki/Precision_and_recall

```bash
               Class     Images     Labels          P          R     mAP@.5 mAP@.5:.95
                 all        128        929      0.577      0.414       0.46      0.279
              person        128        254      0.723      0.531      0.601       0.35
```

For person class:
```
TP = Recall * Labels = 135
FP = TP / Precision - TP = 52
```

**Comment by glenn-jocher** [comment]
@ib124 good news 😃! Your original issue may now be fixed ✅ in PR #5727. This PR explicitly computes TP and FP from the existing Labels, P, and R metrics:

```bash
TP = Recall * Labels
FP = TP / Precision - TP
```

These TP and FP per-class vectors are left in val.py for users to access if they want:
https://github.com/ultralytics/yolov5/blob/36d12a500eae4561d09d4955e1b50b12e57bf6c6/val.py#L240

To receive this update:
- **[Git](https://github.com/ultralytics/yolov5)** – `git pull` from within your `yolov5/` directory or `git clone https://github.com/ultralytics/yolov5` again
- **[PyTorch Hub](https://pytorch.org/hub/ultralytics_yolov5/)** – Force-reload `model = torch.hub.load('ultralytics/yolov5', 'yolov5s', force_reload=True)`
- **[Notebooks](https://github.com/ultralytics/yolov5/blob/master/tutorial.ipynb)** – View updated notebooks  <a href="https://colab.research.google.com/github/ultralytics/yolov5/blob/master/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolov5"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **[Docker](https://hub.docker.com/r/ultralytics/yolov5)** – `sudo docker pull ultralytics/yolov5:latest` to update your image <a href="https://hub.docker.com/r/ultralytics/yolov5"><img src="https://img.shields.io/docker/pulls/ultralytics/yolov5?logo=docker" alt="Docker Pulls"></a>

Thank you for spotting this issue and informing us of the problem. Please let us know if this update resolves the issue for you, and feel free to inform us of any other **issues** you discover or **feature requests** that come to mind. Happy trainings with YOLOv5 🚀!

**Comment by ib124** [comment]
@glenn-jocher This is awesome, thank you! I greatly appreciate this.

**Comment by glenn-jocher** [comment]
@ib124 you're welcome! One thing to note is that these TP and FP values are computed at max-F1 confidence (same as P and R results):
https://github.com/ultralytics/yolov5/blob/7a39803476f8ae55fb25ed93a400a3bba998d5e7/utils/metrics.py#L82-L87
